### PR TITLE
[rollout] fix: derive TQ rollout seeds from global prompt index

### LINKED ISSUE #561: [Bug]: BAGEL FlowGRPO v1 (sync) trainer diverges mid-training where v0 with identical hyperparameters stays stable
## Summary

Running the BAGEL-7B-MoT FlowGRPO PickScore LoRA recipe on the **v1 trainer**
(`verl_omni.trainer.main_diffusion_v1`, `trainer.use_v1=true`,
`trainer.v1.trainer_mode=sync`) diverges mid-training, while the **v0 trainer**
(`verl_omni.trainer.main_diffusion`) with the *same* hyperparameters, data,
seed config, and hardware trains stably for 150+ steps. The v1 run tracks v0
closely for ~30 steps (including a matching first validation score), then
develops an exponentially growing gradient norm from around step 36, and
collapses at steps 59–61: policy KL jumps ~19x, importance ratios fall out of
the trust region, and both training rewards and validation scores collapse.
Nothing crashes — the trainer keeps stepping on a destroyed policy.

This is on the v1 sync migration proposed in #511 (see #389 for the v1
tracking RFC). The divergence is not visible in short smoke tests: the
tiny-random 8-step comparison in #511 looks healthy, and the first real-run
validation at step 30 also looks healthy.

## Environment

- verl-omni: `main` @ `bf048dd` merged with the #511 head (`329d111`)
- Rollout engine: `vllm_omni` (0.28.0rc2.dev68+gded893462), `layered_summon=True`
- 4x H800, single node, `trainer.n_gpus_per_node=4`
- Model: BAGEL-7B-MoT, LoRA rank 64 / alpha 128, `lora_dtype=float32`,
  `param_offload=True`, `optimizer_offload=True`, `model_dtype=bfloat16`
- Data: PickScore parquet (train 25432 / val 2048), `train_batch_size=48`,
  `rollout.n=16`, `ppo_mini_batch_size=24`, `ppo_micro_batch_size_per_gpu=12`
- Optimizer: AdamW, lr 1e-4 (constant), weight_decay 1e-4 — **identical
  settings on both runs**
- `diffusion_loss.clip_ratio=1e-5`, `algorithm.global_std=False`
- Both runs: `save_freq=30`, `test_freq=30`, `val_before_train=False`

Note on the recipe: the v1 script in #511 omits
`actor_rollout_ref.model.lora.merge=True`, which the v0 script sets. Without
it the v1 run produces flat rewards (merged-LoRA weight sync never takes
effect on this model). All symptoms below are with `lora.merge=True` added to
the v1 script so the two recipes are otherwise identical; the only remaining
differences are the entrypoint and `trainer.use_v1=true
trainer.v1.trainer_mode=sync`.

### Phase 1 — steps 1–30: healthy, tracks v0

Step-matched training rewards match v0 within ~0.005 and the first validation
is in line with v0:

| checkpoint | v1 | v0 |
|---|---|---|
| val reward/mean@1 @ step 30 | 0.82661 | 0.83176 |

Two small asymmetries are already visible in this phase:

1. v1's `actor/loss` runs ~2.5–5x v0's from step 1 (e.g. step 1:
   5.0e-06 vs 2.0e-06; step 30: 1.5e-05 vs 6e-07 in similar episodes).
2. v1's `actor/ppo_kl` is sign-alternating noise of magnitude ±1e-5..4e-5,
   while v0's stays at ~1e-7 for its entire 156-step run.

### Phase 2 — ~steps 36–58: exponential gradient growth

`actor/grad_norm` grows roughly exponentially (doubling every ~4–6 steps)
while v0 stays flat:

| step | v1 grad_norm | v1 ppo_kl | v1 ratio_mean | v1 reward | v0 ppo_kl | v0 reward |
|---|---|---|---|---|---|---|
| 36 | 7.6e-03 | 7.6e-04 | 0.9992 | 0.8320 | -2.3e-07 | 0.8442 |
| 40 | 1.5e-02 | 1.2e-03 | 0.9988 | 0.8127 | 9.5e-07 | 0.8356 |
| 45 | 1.9e-02 | 2.4e-03 | 0.9976 | 0.8119 | -5.9e-07 | 0.8339 |
| 50 | 2.4e-02 | 5.3e-04 | 0.9995 | 0.8264 | 6.1e-07 | 0.8396 |
| 55 | 5.6e-02 | 4.8e-03 | 0.9958 | 0.8315 | -7.6e-07 | 0.8423 |
| 57 | 1.6e-01 | 2.0e-03 | 0.9983 | 0.8224 | 1.4e-06 | 0.8403 |
| 58 | 1.2e-01 | 7.8e-03 | 0.9938 | 0.8122 | 3.5e-06 | 0.8431 |

### Phase 3 — steps 59–61: collapse

| step | v1 grad_norm | v1 ppo_kl | v1 ratio_mean | pg_clipfrac_lower | v1 reward |
|---|---|---|---|---|---|
| 59 | 3.2e-01 | 5.4e-02 | 0.9586 | 0.457 | 0.7672 |
| 60 | 2.4e-01 | 3.1e-02 | 0.9795 | 0.357 | 0.6820 |
| 61 | 3.8e-01 | 5.9e-02 | 0.9574 | 0.426 | 0.6985 |

Validation at step 60:

| | v1 | v0 |
|---|---|---|
| val reward/mean@1 @ step 60 | **0.74792** | 0.84020 |

v0, same data/seed/hardware, ran 156 steps with `ratio_mean ≡ 1.0000`,
`ppo_kl ~ 1e-7`, `grad_norm ~ 1e-4` throughout, and val rising
0.832 → 0.855.

Infra-wise the v1 run was healthy at all times: steady ~480 s/step, no OOM,
no preemptions, no errors; `update_weights` executes every step (~12–13 s,
same as v0); the collapse is purely in the training dynamics.

## What we already checked (all identical between v0 and v1)

- Merged-LoRA weight sync round-trip: frozen base weights in the checkpoints
  are **bit-identical** to the pretrained base (max |delta| = 0.0) in both
  trainers; sync executes once per step in both.
- `old_log_prob` is recomputed after the latest weight sync and before the
  actor update in both; `ratio_mean` starts each step at exactly 1.0000 in
  both.
- Optimizer steps per global step: 2 in both (same mini-batch arithmetic).
- Advantage computation: same `group_size=16`, `zero_std_ratio=0`, matching
  advantage mean/max/min series.
- Full config dumps are semantically identical (whitespace-normalized diff
  shows only paths/experiment names).
- On the tiny-random smoke (`tests/special_e2e/run_flowgrpo_bagel_pickscore.sh`
  vs `..._v1.sh`) with LoRA init seeded to the same values, the first-update
  gradient matches within ~1.2x and the recomputed old log probs are
  bitwise-equal in both trainers (ppo_kl = 0.0).
- One reproducibility observation: LoRA init is drawn from the global
  (unseeded) RNG per worker, so early-run loss/grad scale varies run to run
  in both trainers; the step-1 loss ratio between any two runs (v0 vs v0
  included) can vary by several x.

## Reproduction

```bash
# v0: stable (156+ steps, flat grad_norm ~1e-4, val rising)
bash examples/flowgrpo_trainer/bagel/run_bagel_pickscore_lora.sh

# v1: same data/config, diverges as described above
# (with actor_rollout_ref.model.lora.merge=True added, see note)
bash examples/flowgrpo_trainer/bagel/run_bagel_pickscore_lora_v1.sh
```

Timeline at ~480 s/step: steps 1–30 healthy (first mismatch signals appear
in actor/loss scale and ppo_kl noise), grad_norm growth becomes clear by
step ~36–45, collapse at ~step 59–61. A controlled rerun with additional
instrumentation is in progress; results will be added as comments.

### PR DESCRIPTION
**Pull request by OnePunchMonk**
## Summary

Fixes #561.

The v1 TQ agent-loop worker derived per-rollout RNG seeds from the **chunk-local** prompt index (`sample_index`, 0-11 per worker) instead of the **global** prompt index (0-47), unlike the v0 path. With 48 prompts chunked 12/12/12/12 across 4 workers, this caused 4 different prompts (one per worker, at the same chunk-local position) to receive identical seeds every step, producing identical noise realizations across their rollout groups. This is a structural, deterministic corruption of rollout noise that inflates policy-gradient variance and destabilizes GRPO training.

Fix: `verl_omni/agent_loop/diffusion_agent_loop_tq.py` now derives `sample_index` from the batch's global `index` field (already available in scope) instead of the loop-local `i`, matching v0's convention in `diffusion_agent_loop.py` / `maybe_per_rollout_seeds`.

## AI assistance disclosure

This PR was AI-assisted: **Claude** (Claude Code) diagnosed/scoped the fix from the issue and orchestrated the work, and **OpenAI Codex** wrote the production diff and the regression test. Per this repo's contribution policy, a human submitter (me) will review every changed line and confirm the reasoning before this leaves draft status.

**This PR stays in draft until I (the human submitter) have personally reviewed every changed line end-to-end and am prepared to defend the change. I will mark it "ready for review" only after that pass — not before.**

## Test plan

Ran the regression test (and the two other `seed`/`tq`-matched tests in the same file) on a Modal GPU container (A10G, CUDA 13.0, Ubuntu 22.04), installing the actual pinned dependency stack per `docs/start/install.md`:
- `uv pip install -e ".[gpu]" --torch-backend=auto`
- `uv pip install "vllm-omni @ git+https://github.com/vllm-project/vllm-omni.git@ded8934626aaad1a3e816c3a1d9d742efc012d93"` (the exact commit pinned in `.github/vllm_omni_pin.txt`, and the same commit cited in the issue's root-cause analysis)
- `uv pip install -e ".[train,dev]"`

Command:
```
.venv/bin/python -m pytest -q --asyncio-mode=auto \
  tests/agent_loop/test_diffusion_agent_loop_correctness_on_cpu.py -k "seed or tq" -v
```

Result:
```
tests/agent_loop/test_diffusion_agent_loop_correctness_on_cpu.py ...     [100%]
3 passed, 18 deselected, 20 warnings in 34.43s
```

The new test, `test_tq_rollout_seeds_are_global_across_worker_chunks`, simulates the exact 48-prompt / 4-worker / 12-per-chunk scenario from #561, asserts all `48 * 16 = 768` derived seeds are distinct, and asserts they match `maybe_per_rollout_seeds` (v0's derivation) exactly. It fails against the pre-fix code (chunk-local seeding collapses to 192 distinct values) and passes after the fix.

Earlier attempts to run this on CPU-only environments failed for environment reasons unrelated to the fix (missing `vllm_omni`, then a missing CUDA runtime lib for `torchcodec`'s import chain) — the GPU container above was used specifically to get a clean, representative result against the repo's real dependency graph.

I have not run the full test suite (blocked in some earlier CPU-only attempts by an unrelated `datasets`/`pyarrow` incompatibility, not investigated further since out of scope for this fix).

### CODE REVIEW

**Comment by OnePunchMonk** [comment]
Self-review: This should fix seed collision; will check for long-term training collapse next.
